daft/runners/pyrunner.py

```diff
@@ -279,41 +285,58 @@ def _physical_plan_to_partitions(
                             # update progress bar
                             pbar.mark_task_start(next_step)
 
+                            with self._resource_accounting_lock:
+                                # Update resource accounting
+                                self._available_bytes_memory -= next_step.resource_request.memory_bytes or 0
+                                self._available_cpus -= next_step.resource_request.num_cpus or 0.0
+                                self._available_gpus -= next_step.resource_request.num_gpus or 0.0
+                                assert (
+                                    self._available_bytes_memory >= 0
+                                ), "Available bytes in memory should not go below 0. This indicates a scheduler bug."
+                                assert (
+                                    self._available_cpus >= 0
+                                ), "Available CPUs should not go below 0. This indicates a scheduler bug."
+                                assert (
+                                    self._available_gpus >= 0
+                                ), "Available GPUs should not go below 0. This indicates a scheduler bug."
+
```

Comment:
Also this resource subtraction should happen in the same critical section as when we check for available resources. Otherwise we can get a race condition where multiple threads can confirm that they have the necessary resources before any thread grabs resources, even though there are not enough resources for every thread.

E.g. let's say there's only 1 cpu available, we can encounter the following scenario:
| execution 1            | execution 2            |
|------------------------|------------------------|
| _can_admit_task = true |                        |
|                        | _can_admit_task = true |
| subtract 1 cpu (remaining cpu = 0), ok |                        |
|                        | subtract 1 cpu (remaining cpu = -1), **assertion fails** |